# PES University, Bangalore
Established under Karnataka Act No. 16 of 2013

UE20CS312 - Data Analytics - Worksheet 5
Course instructor: Gowri Srinivasa, Professor Dept. of CSE, PES University

Designed by Yashas Kadambi, Dept. of CSE - yashasks@pesu.pes.edu

Submitted by: Renita Kurian, PES1UG20CS331

# Markov Chains and AB Testing

### Prerequisites
- Revise the following concepts
    - Markov Chains
    - Markov Chains with Absorbing states
    - Calculation of eventual probability of aborbtion
    - A/B Testing
- Install the following software
    - pandas
    - numpy

### Points
The problems in this worksheet are for a total of 10 points with each problem having a different weightage.
- Problem 1: 2 points
- Problem 2: 4 points
- Problem 3: 4 points

# Scenario 2

Its a freezing day in New york, Commisioner Wench has sent a report to Captain Holt that the 99th precinct has much lower reported crimes compared to other precincts. Upon Analysis Captain Holt decides to add feedback unit along with the 4 major units to analyse this descripency. All the units are mentioned below

1.   Major Crimes
2.   Traffic
3.   General crimes
4.   Feedback
5.   Theft

---
---
<br>

The initial probablity of a person going to a particular unit on a particular day is given as follows


<br>

Major Crimes | Traffic | General crimes | Feedback | Theft
:----------: | :-----: | :------------: | :--------: | :---:
0.3 | 0.4 | 0.1 | 0.15 | 0.05 
<br>


To measure how many people will go to the feedback unit, the personel files of all employees are give to the **_Move-o-Tron 99_** and it gives us the following matrix which shows us the probability of people moving from one unit to another on a particular day . It is known that the **_Move-o-Tron 99_** alwasy outputs matices which follow a first order Markov chain. 

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.666|0\.31|0\.0|0\.022|
|Traffic|0\.466|0\.102|0\.222|0\.0|0\.21|
|General crimes|0\.022|0\.11|0\.502|0\.0|0\.366|
|Feedback|0\.0|0\.0|0\.0|1\.0|0\.0|
|Theft|0\.11|0\.122|0\.066|0\.0|0\.702|

As the people of New York are smart the will learn where all the units are present and hence the next days (day 1) distribution will be the distribution present at the end of the current day (day 0). Captain holt want to check if the matrix given by the **_Move-o-Tron_** can be used to model the footfall.



## Problem 1 (2 points)

1. What technique can be used to model the probability of people going to the correct unit to report thier crime after N days? (0 points)
2. Is the chain irreducible? Justify (0.5 point)
3. What will be the intital probability of a person going to a particular unit after 1 day, 2 days, 10 days, 1000 days and 1001 days. (1 point)
 
 Hint: Use the  Chapman−Kolmogorov relationship

     ```python
    # C = A.B
    matrix_C = np.dot(matrix_A, matrix_B)

    # C = A.(B^4) can be replaced by
    matrix_C = matrix_A
    for _ in range(4):
        matrix_C = np.dot(matrix_C, matrix_B)
    ```

4. What can you say about the markov chain from state of intital probability of a person going to a particular unit after 1000 and 1001 days? (0.5 points)

In [1]:
# Importing Libraries
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [2]:
# encoding the probabilities as a numpy array
trans_array = np.array([
    [0.002, 0.666, 0.31, 0, 0.022],
    [0.466, 0.102, 0.222, 0, 0.21],
    [0.022, 0.11, 0.502, 0, 0.366],
    [0, 0, 0, 1, 0],
    [0.11, 0.122, 0.066, 0, 0.702]
])
# ensures that the probabilities sum to 1
assert(trans_array[0].sum() == 1.0)
assert(trans_array[1].sum() == 1.0)
assert(trans_array[2].sum() == 1.0)
assert(trans_array[3].sum() == 1.0)
assert(trans_array[4].sum() == 1.0)



# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)

### Answers
Q1. Markov Chains is a suitable method  
Q2. No, the chain is not irreducible. The Markov Chain is reducible as feedback unit is not communicable.

In [3]:
# 1 day
s1 = np.dot(state, trans_array)
print(s1[0])

[0.1947 0.2577 0.2353 0.15   0.1623]


In [4]:
# 2 days
s2 = np.dot(s1, trans_array)
print(s2[0])

[0.1435072 0.2016392 0.2463988 0.15      0.2584548]


In [5]:
# 10 days
s10 = state
for i in range(10):
  s10 = np.dot(s10, trans_array)
print(s10[0])

[0.12139467 0.16406836 0.1973962  0.15       0.36714077]


In [6]:
# 1000 days
s1000 = state
for i in range(1000):
  s1000 = np.dot(s1000, trans_array)
print(s1000[0])

[0.1214373  0.16411091 0.19739717 0.15       0.36705462]


In [7]:
# 1001 days
s1001 = np.dot(s1000, trans_array)
print(s1001[0])

[0.1214373  0.16411091 0.19739717 0.15       0.36705462]


Q3.   
1 day - [0.1947 0.2577 0.2353 0.15   0.1623]  
2 days -  [0.1435072 0.2016392 0.2463988 0.15 0.2584548]  
10 days - [0.12139467 0.16406836 0.1973962  0.15  0.36714077]  
1000 days - [0.1214373  0.16411091 0.19739717 0.15 0.36705462]  
1001 days - [0.1214373  0.16411091 0.19739717 0.15  0.36705462]  

Q4. s1000 = s1001, it is observed that there is no change in probabilities (Stationary)

---

After analysing the model Captain holt calls the squad and educates them to ask people to give feedbacks. And the details of the squad are given to the **_Move-o-Tron 99_**. After reanalyising the report the **_Move-o-Tron 99_** gave out a new Matrix, which is shown below

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.666|0\.01|0\.02|0\.302|
|Traffic|0\.466|0\.102|0\.02|0\.032|0\.38|
|General crimes|0\.0|0\.0|1\.0|0\.0|0\.0|
|Feedback|0\.0|0\.0|0\.0|1\.0|0\.0|
|Theft|0\.11|0\.122|0\.066|0\.004|0\.698|

<br>
Considering the new report the model has to be re-evaluated. The initial probablity of a person going to a particular unit on a particular day remains the same.
<br>





## Problem 2 (4 points)


1. Is the chain irreducible? Justify (0.5 point)
2. What will be the intital probability of a person going to a particular unit after 1 day, 2 days, 10 days, 1000 days and 1001 days. (1 point)
 
 Hint: Use the  Chapman−Kolmogorov relationship

3. What can you say about the markov chain from state of intital probability of a person going to a particular unit after 1000 and 1001 days? (0.5 points)


4. Summer Edgecombe is Confidential Informant (CI) to the Officer Kimbal Cho and comes in every day to the police station. If on the first day she goes to the Major crimes Unit what will be the probability that she gives a feedback? (2 points)

    ```python
    # np.delete()
    # https://note.nkmk.me/en/python-numpy-delete/#:~:text=Using%20the%20NumPy%20function%20np,from%20the%20NumPy%20array%20ndarray%20.&text=Specify%20the%20axis%20(dimension)%20and,a%20slice%20or%20a%20list.

    print(a)
    # [[ 0  1  2  3]
    #  [ 4  5  6  7]
    #  [ 8  9 10 11]]

    print(np.delete(a, 1, 0))
    # [[ 0  1  2  3]
    #  [ 8  9 10 11]]

    print(np.delete(a, 1, 1))
    # [[ 0  2  3]
    #  [ 4  6  7]
    #  [ 8 10 11]]

    # Deleting multiple rows or columns
    print(np.delete(a, [0, 3], 1))
    # [[ 1  2]
    #  [ 5  6]
    #  [ 9 10]]

    # Deleting rows and columns
    print(np.delete(np.delete(a, 1, 0), 1, 1))
    # [[ 0  2  3]
    #  [ 8 10 11]]

    # matrix multiplication or cross pdt C = A*B
    matrix_C = matrix_A @ matrix_B # matrix_C = np.matmul(matrix_A, matrix_B)

    ```

In [8]:
# encoding the probabilities as a numpy array
trans_array = np.array([
    [0.002, 0.666, 0.01, 0.020, 0.302],
    [0.466, 0.102, 0.02, 0.032, 0.38],
    [0.0, 0.0, 1, 0.0, 0.0],
    [0, 0, 0, 1, 0],
    [0.11, 0.122, 0.066, 0.004, 0.698]
])

# ensures that the probabilities sum to 1
assert(trans_array[0].sum() == 1.0)
assert(trans_array[1].sum() == 1.0)
assert(trans_array[2].sum() == 1.0)
assert(trans_array[3].sum() == 1.0)
assert(trans_array[4].sum() == 1.0)

# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)


### Answers  
Q1. No, the Markov Chain is not irreducible.

In [9]:
#Q2
# 1 day
s1 = np.dot(state, trans_array)
print(s1[0])

[0.1925 0.2467 0.1143 0.169  0.2775]


In [10]:
# 2 days
s2 = np.dot(s1, trans_array)
print(s2[0])

[0.1458722 0.1872234 0.139474  0.1818544 0.345576 ]


In [11]:
# 10 days
s10 = state
for i in range(10):
  s10 = np.dot(s10, trans_array)
print(s10[0])

[0.07693332 0.09688565 0.3362719  0.24330485 0.24660428]


In [12]:
# 1000 days
s1000 = state
for i in range(1000):
  s1000 = np.dot(s1000, trans_array)
print(s1000[0])

[8.97378472e-28 1.13004326e-27 6.60595331e-01 3.39404669e-01
 2.87688168e-27]


In [13]:
# 1001 days
s1001 = np.dot(s1000, trans_array)
print(s1001[0])

[8.44851901e-28 1.06389804e-27 6.60595331e-01 3.39404669e-01
 2.70848815e-27]


Q2.   
1 day - [0.1925 0.2467 0.1143 0.169  0.2775]  
2 days - [0.1458722 0.1872234 0.139474  0.1818544 0.345576 ]  
10 days - [0.07693332 0.09688565 0.3362719  0.24330485 0.24660428]  
1000 days - [8.97378472e-28 1.13004326e-27 6.60595331e-01 3.39404669e-01
 2.87688168e-27]  
1001 - [8.44851901e-28 1.06389804e-27 6.60595331e-01 3.39404669e-01
 2.70848815e-27]  
 
Q3. There is slight change between s1000 and s1001. (Not stationary distribution)

In [14]:
#Q4
x = np.delete(np.delete(trans_array, [2,3], 0), [2,3], 1)
print(x)

[[0.002 0.666 0.302]
 [0.466 0.102 0.38 ]
 [0.11  0.122 0.698]]


In [15]:
y = np.delete(np.delete(trans_array, [2,3], 0), [0,1,4], 1)
print(y)

[[0.01  0.02 ]
 [0.02  0.032]
 [0.066 0.004]]


In [16]:
identity = np.identity(3)
z = np.linalg.inv(identity - x)
print(z)

[[ 4.0279365   4.26333958  9.39240352]
 [ 3.27006043  4.80437252  9.31529737]
 [ 2.78814698  3.49371126 10.49546579]]


In [17]:
# Eventual Absorption
#print(z@y)
print((z@y)[1][1])

0.25640231868803715


Probability that she gives feedback - 0.256

## Problem 3 (4 points)

It seems that there is a bug in **_Move-o-Tron 99_** which makes general crimes and feedback units as absorbing states. After updating the software of **_Move-o-Tron 99_**, Captain Holt wants to find out the effect that Amy Santiago has on the probability of a person giving feedback. So one matrix is generated including Santiago and the other one without. 

Matrix 1 (With Santiago)

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.002|0\.232|0\.31|0\.434|0\.022|
|Traffic|0\.426|0\.102|0\.222|0\.04|0\.21|
|General crimes|0\.03|0\.11|0\.2|0\.294|0\.366|
|Feedback|0\.003|0\.176|0\.321|0\.3|0\.2|
|Theft|0\.11|0\.422|0\.166|0\.1|0\.202|

Matrix 2 (Without Santiago)

| |Major Crimes|Traffic|General crimes|Feedback|Theft|
|---|---|---|---|---|---|
|Major Crimes|0\.11|0\.222|0\.092|0\.374|0\.202|
|Traffic|0\.03|0\.11|0\.2|0\.294|0\.366|
|General crimes|0\.002|0\.232|0\.31|0\.434|0\.022|
|Feedback|0\.466|0\.102|0\.02|0\.032|0\.38|
|Theft|0\.003|0\.176|0\.321|0\.3|0\.2|

<br>

1. How can you find out the effect that Santiago has on the probability of feedback? (1 point)

2. What effect does Santiago have one the probability of getting feedback? (1 point)

    Note: The initial probablity of a person going to a particular unit on a particular day remains the same

3. Name the test Captain Holt is performing. (0.5 points)

Lina Ginetti reports to Captain Holt that the there two kinds of days in the precicnt _"There are normal days and then there are days where workflow is dismal with a tiny dash of pathetic."_. Captain Holt decided to sample the initial probablity of a person going to a particular unit on a good day and a bad day.

4. Without the information about these inital probabilities, can you tell if there is any difference in the probability of getting a feedback? Explain. (1.5 points)



In [18]:
# encoding the probabilities as a numpy array
# With Santiago
trans_array_with_amy = np.array([
    [0.002, 0.232, 0.31, 0.434, 0.022],
    [0.426, 0.102, 0.222, 0.04, 0.21],
    [0.03, 0.11, 0.20, 0.294, 0.366],
    [0.003, 0.176, 0.321, 0.3, 0.2],
    [0.11, 0.422, 0.166, 0.1, 0.202]
])

# Without Santiago
trans_array_without_amy = np.array([
    [0.11, 0.222, 0.092, 0.374, 0.202],
    [0.03, 0.11, 0.20, 0.294, 0.366],
    [0.002, 0.232, 0.31, 0.434, 0.022],
    [0.466, 0.102, 0.02, 0.032, 0.38],
    [0.003, 0.176, 0.321, 0.3, 0.2]
])

# ensures that the probabilities sum to 1
assert(trans_array_with_amy[0].sum() == 1.0)
assert(trans_array_with_amy[1].sum() == 1.0)
assert(trans_array_with_amy[2].sum() == 1.0)
assert(trans_array_with_amy[3].sum() == 1.0)
assert(trans_array_with_amy[4].sum() == 1.0)

assert(trans_array_without_amy[0].sum() == 1.0)
assert(trans_array_without_amy[1].sum() == 1.0)
assert(trans_array_without_amy[2].sum() == 1.0)
assert(trans_array_without_amy[3].sum() == 1.0)
assert(trans_array_without_amy[4].sum() == 1.0)

# encoding the initial probability of as a numpy array
state = np.array([[0.3, 0.4, 0.1, 0.15, 0.05]])
assert(state[0].sum() == 1.0)


### Answers
Q1. Compare probabilities of stationary distribution for both transition matrices.

In [19]:
# with
s1000x = state
for i in range(1000):
  s1000x = np.dot(s1000x, trans_array_with_amy)
print(s1000x[0])
s1001x = np.dot(s1000x, trans_array_with_amy)
print(s1001x[0])

[0.12001975 0.20607988 0.23649985 0.21700347 0.22039705]
[0.12001975 0.20607988 0.23649985 0.21700347 0.22039705]


In [20]:
# without
s1000y = state
for i in range(1000):
  s1000y = np.dot(s1000y, trans_array_without_amy)
print(s1000y[0])
s1001y = np.dot(s1000y, trans_array_without_amy)
print(s1001y[0])

[0.14495178 0.16283362 0.18658671 0.26400004 0.24162786]
[0.14495178 0.16283362 0.18658671 0.26400004 0.24162786]


Q2.   
With Amy - 0.217  
Without Amy - 0.264  
with Amy < without Amy (s1000x < s1000y). Hence, getting a feedback with Amy working is less.  
Q3. AB Testing  
Q4. Yes. Stationary distribution does not depend on initial states.